### PR TITLE
add functionality to get comments by article Id

### PR DESCRIPTION
**Pull request by jayascript1**

### CODE REVIEW

File: __tests__/app.test.js [hunk 1]
```diff
@@ -102,4 +102,52 @@ describe("GET /api/articles", () => {
...
   });
 
+  describe("GET /api/articles/:article_id/comments", () => {
+    it("responds with an array of comments for the given article_id with the required properties", () => {
+      return request(app)
+        .get("/api/articles/1/comments")
+        .expect(200)
+        .then((res) => {
+          const comments = res.body;
+          comments.forEach((comment) => {
```
Comment (callumu21) [inline __tests__/app.test.js]: When using .forEach, we also have to assert against the length of the array. Otherwise, an empty array would automatically pass this test and give us a false positive

File: __tests__/app.test.js [hunk 2]
```diff
@@ -102,4 +102,52 @@ describe("GET /api/articles", () => {
...
+            expect(comment).toMatchObject({
+              comment_id: expect.any(Number),
+              votes: expect.any(Number),
+              created_at: expect.any(String),
+              author: expect.any(String),
+              body: expect.any(String),
+              article_id: expect.any(Number),
```
Comment (callumu21) [inline __tests__/app.test.js]: For article_id, they all should have the same value, rather than just any number

File: __tests__/app.test.js [hunk 3]
```diff
@@ -102,4 +102,52 @@ describe("GET /api/articles", () => {
...
+        });
+    });
+    it("serves the most recent comments first", () => {
+      return request(app)
+        .get("/api/articles/1/comments")
+        .expect(200)
+        .then((res) => {
+          const comments = res.body;
+          expect(comments.length).toBeGreaterThan(0);
+          const timestamps = comments.map((comment) =>
```
Comment (callumu21) [inline __tests__/app.test.js]: You create these timestamps, but don't appear to be doing anything with them. My advice would be to use jest-sorted here to check if the created_at value is in a specific order

File: __tests__/app.test.js [hunk 4]
```diff
@@ -102,4 +102,52 @@ describe("GET /api/articles", () => {
...
+              comment_id: expect.any(Number),
+              votes: expect.any(Number),
+              created_at: expect.any(String),
+              author: expect.any(String),
+              body: expect.any(String),
+              article_id: expect.any(Number),
+            });
+          });
+        });
+    });
```
Comment (callumu21) [inline __tests__/app.test.js]: We should also test that an article with no comments returns an empty array and a successful status code

File: __tests__/app.test.js [hunk 5]
```diff
@@ -102,4 +102,52 @@ describe("GET /api/articles", () => {
...
+            new Date(comment.created_at).getTime()
+          );
+        });
+    });
+    it("handles 404 error when the article_id does not exist", () => {
+      return request(app)
+        .get("/api/articles/999/comments")
+        .then((res) => {
+          expect(404);
+          expect(res.body.message).toEqual("Article not found");
```
Comment (callumu21) [inline __tests__/app.test.js]: Nice error messages. Useful for the client, but it doesn't expose any details of the BE

File: controllers/comments.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
+const { fetchCommentsByArticleId } = require('../models/comments');
+
+exports.getCommentsByArticleId = (req, res, next) => {
+  const articleId = req.params.article_id;
+  if (isNaN(articleId)) {
```
Comment (callumu21) [inline controllers/comments.js:6]: Nice logic to create a custom error message, but remember that PSQL will give us an error message for this automatically without having to add extra logic

File: controllers/comments.js [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+    return res.status(400).json({ message: 'Invalid article ID' });
+  }
+  fetchCommentsByArticleId(articleId)
+    .then((comment) => {
+      if (comment.length === 0) {
+        return res.status(404).send({ message: 'Article not found'})
```
Comment (callumu21) [inline controllers/comments.js]: This logic should be handled in the model, throwing a rejected promise that is handled by your error handling middleware functions.

File: controllers/comments.js [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+      }
+      return res.status(200).json(comment);
```
Comment (callumu21) [inline controllers/comments.js]: We should be returning an object with a key of comments here that contains the array of comments as the value

**Review by callumu21 — CHANGES_REQUESTED** [review]
Nice start, but a few changes are needed for this to be approved:

1. When using .forEach, we need to assert against the length of the array as well to avoid a false positive. Currently your first test doesn't check that all of the comments are being returned. We should also check that the article_id is correct for each comment, rather than just any number.
2. We should have a test for an empty array being returned when an article has no comments
3. It doesn't look like your test that the comments are in reverse order is working correctly. I would look at jest-sorted to make it easier.
4. We should return the comments on a key of comments on an object, rather than just the array directly.

Once you've incorporated these changes, please do resubmit the PR and we'll take another look 👀

File: __tests__/app.test.js [hunk 9]
```diff
@@ -90,16 +90,81 @@ describe("GET /api/articles", () => {
...
+        const comments = res.body.comments
+        let commentCount = 0;
+        let i = 0;
+        while (i < testData.commentData.length) {
+          if ((testData.commentData[i].article_id === articleId)) {
+            commentCount++;
+          }
+          i++;
+        }
+        expect(comments.length).toBe(commentCount);
```
Comment (callumu21) [inline __tests__/app.test.js]: I enjoy that you've tried to handle this programatically, but it's currently a loop that checks all returned comments have the correct article_id, but not that all comments for a specific article_id have been returned.

You can just hardcode the length value by looking at the test data you have access to (e.g. expect(comments.length).toBe(14))

File: __tests__/app.test.js [hunk 10]
```diff
@@ -90,16 +90,81 @@ describe("GET /api/articles", () => {
...
+        });
+      });
+  });
+  it("serves the most recent comments first", () => {
+    return request(app)
+      .get("/api/articles/1/comments")
+      .expect(200)
+      .then((res) => {
+        const comments = res.body.comments;
+        expect(comments).toBeSortedBy('created_at', {descending: true})
```
Comment (callumu21) [inline __tests__/app.test.js:135]: Great use of jest-sorted here!

File: controllers/comments.js [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+  fetchArticleById(articleId)
+    .then((article) => {
+      if (!article) {
+        const err = new Error('Article not found')
+        err.status = 404
+        throw err
+      }
+      fetchCommentsByArticleId(articleId)
+        .then((comments) => {
+          if (comments.length === 0) {
```
Comment (callumu21) [inline controllers/comments.js]: These lines are doing the same thing, so you can just have line 23 and remove the conditional logic 👍

**Review by callumu21 — APPROVED** [review]
Nice job implementing those changes! I think you've slightly overcomplicated finding the length of the comments array by using a loop. Please refactor this with the number of comments hardcoded based on the test data, as that's the only way you'll know that all of them have come back.

Once you've added that in, this should be good to merge 👍
